### PR TITLE
Add support for V2 project handles

### PR DESCRIPTION
**Pull request by peripheralist**
### Form
Adds form for setting V2 project handle, accessible through the newDeploy modal, or through the project settings modal. The form includes 2 separate transactions:
- Setting the ENS name ("reverse record") for the project in the JBProjectHandles contract
- Setting the text record on that ENS name with a value of the project's ID

### Router
Routing adjusted to accommodate handles in v2 project URLs
- No existing routes have been changed
- V2 projects with handles can now found at juicebox.money/#/@handle

## Screenshots or screen recordings

Prompt to set handle from project header:
image

Flow for setting handle from newDeploy menu. All values read from contract calls will update automatically once transactions succeed, no need to refresh between steps. Note in last step once the handle is set, the handle appears on the project page and step 1 of the newDeploy modal menu is grayed out.
Screen Shot 2022-06-17 at 1 17 22 PM
Screen Shot 2022-06-17 at 1 17 40 PM
Screen Shot 2022-06-17 at 1 18 12 PM
Screen Shot 2022-06-17 at 1 18 53 PM

Search results:
Screen Shot 2022-06-17 at 1 20 07 PM

Edit project settings modal:
image

UPDATE—copy changes to handle modal:
image

## Acceptance checklist

- [x] I have evaluated the
      [Approval Guidelines](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#approval-guidelines)
      for this PR.
- [x] I have tested this PR in
      [all supported browsers](https://github.com/jbx-protocol/juice-interface/blob/main/CONTRIBUTING.md#supported-browsers).

### CODE REVIEW

**Review by wraeth-eth — APPROVED** [review]
Looks good! Made a few suggestions, but nothing major

File: src/hooks/v2/contractReader/ProjectHandleENSTextRecord.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+export function useProjectHandleENSTextRecord(ensName: string | undefined) {
+  return useV2ContractReader<number>({
+    contract: V2ContractName.PublicResolver,
+    functionName: 'text',
+    args: ensName ? [ensName, projectHandleENSTextRecordKey] : null,
+    formatter: (val: string) => {
+      let result = 0
+
+      try {
```
[the remark this answers is not in the document]
Comment (peripheralist) [reply src/hooks/v2/contractReader/ProjectHandleENSTextRecord.ts:19]: thanks @wraeth-eth for those little catches 🙏

File: src/components/shared/Project/ProjectHeader/index.tsx [hunk 2]
```diff
@@ -98,23 +103,34 @@ export default function ProjectHeader({
...
                 style={{
                   color: colors.text.secondary,
-                  marginRight: spacing,
                   fontWeight: 600,
                 }}
               >
                 @{handle}
               </span>
-            )}
+            ) : onClickSetHandle ? (
```
Comment (peripheralist) [inline src/components/shared/Project/ProjectHeader/index.tsx:120]: if no handle is set and owner wallet is connected, show button prompt to set handle

File: src/hooks/IsUserAddress.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { NetworkContext } from 'contexts/networkContext'
+import { useContext } from 'react'
+
+export function useIsUserAddress(address: string | undefined) {
```
Comment (peripheralist) [inline src/hooks/IsUserAddress.ts:4]: helper util to avoid duplicate logic

File: src/utils/routes.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { BigNumber, BigNumberish } from '@ethersproject/bignumber'
+
+export const v2ProjectRoute = ({
```
Comment (peripheralist) [inline src/utils/routes.ts:3]: helper util to enforce consistency

File: src/components/v2/V2Project/NewDeployModal.tsx [hunk 5]
```diff
@@ -61,6 +66,25 @@ export default function NewDeployModal({
       <div>
         <RichButton
           prefix="1"
+          heading={<Trans>Set a project handle (optional)</Trans>}
```
Comment (peripheralist) [inline src/components/v2/V2Project/NewDeployModal.tsx:69]: Adds "set handle" step to newDeploy modal, which opens drawer with form to configure project handle

File: src/components/v2/V2Project/V2ProjectReconfigureModal/index.tsx [hunk 6]
```diff
@@ -160,13 +163,24 @@ export default function V2ProjectReconfigureModal({
                 Changes to project details will take effect immediately.
               </Trans>
             </p>
-            <ReconfigureButton
```
Comment (peripheralist) [inline src/components/v2/V2Project/V2ProjectReconfigureModal/index.tsx:163]: only layout changes to this component—adding new section for configuring project handle

File: src/components/v2/shared/V2ProjectTokenBalance/V2ProjectTokenBalance.tsx [hunk 7]
```diff
@@ -10,14 +10,14 @@ import { tokenSymbolText } from 'utils/tokenSymbolText'
 
 export const V2ProjectTokenBalance = ({
   projectId,
+  handle,
   style,
   precision,
-  hideHandle,
```
Comment (peripheralist) [inline src/components/v2/shared/V2ProjectTokenBalance/V2ProjectTokenBalance.tsx:15]: unused property. removed for convenience

File: src/hooks/v2/V2ContractLoader.ts [hunk 8]
```diff
@@ -73,12 +78,40 @@ const loadContract = async (
...
 ): Promise<Contract | undefined> => {
-  const contractOverride = CONTRACT_ABI_OVERRIDES[contractName]
-  const version = contractOverride?.version ?? 'latest'
-  const filename = contractOverride?.filename ?? contractName
-
-  const contract = await import(
-    `@jbx-protocol/contracts-v2-${version}/deployments/${network}/${filename}.json`
-  )
-  return new Contract(contract.address, contract.abi, signerOrProvider)
+  let contractJson: { abi: ContractInterface; address: string } | undefined =
```
Comment (peripheralist) [inline src/hooks/v2/V2ContractLoader.ts:81]: new pattern for importing contracts from multiple sources.

note: we aren't able to import ENS contracts from packages, so those contract artifacts have been added to src/constants for now

**Review by aeolianeth — APPROVED** [review]

**Comment by aeolianeth** [comment]
@peripheralist massive massive effort, thanks for persisting with this. I think the `@` is a great solution.

I had a few minor suggestions but in the interest of getting this shipped, didn't feel compelled to comment. LFG. ✅

**Comment by peripheralist** [comment]
thanks @aeolianeth ! good catches, fixed those points

**Pull request merged**